src/ring_theory/trace.lean

```diff
@@ -0,0 +1,126 @@
+/-
+Copyright (c) 2020 Anne Baanen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Anne Baanen
+-/
+
+import linear_algebra.bilinear_form
+import ring_theory.power_basis
+
+/-!
+# Trace for (finite) ring extensions.
+
+Suppose we have an `R`-algebra `S` with a finite basis. For each `s : S`,
+the trace of the linear map given by multiplying by `s` gives information about
+the roots of the minimal polynomial of `s` over `R`.
+
+## Implementation notes
+
+Typically, the trace is defined specifically for finite field extensions.
+The definition is as general as possible and the assumption that we have
+fields or that the extension is finite is added to the lemmas as needed.
+
+We only define the trace for left multiplication (`algebra.left_mul_matrix`,
+i.e. `algebra.lmul_left`).
+For now, the definitions assume `S` is commutative, so the choice doesn't matter anyway.
+
+## References
+
+ * https://en.wikipedia.org/wiki/Field_trace
+
+-/
+
+universes u v w
+
+variables {R S T : Type*} [comm_ring R] [comm_ring S] [comm_ring T]
+variables {A: Type*} [integral_domain A] [algebra A S]
```

Comment:
```suggestion
variables {A : Type*} [integral_domain A] [algebra A S]
```
although it appears you don't actually use `A` anyway?

Reply:
Not yet. Should I just delete it for now and re-add it in the commit that needs it?

Reply:
I think that would be good, otherwise its hard to review whether you have the right typeclass